// ngx-search-pipe/fesm2020/ngx-search-pipe.mjs
import * as i0 from '@angular/core';
import { Pipe, NgModule } from '@angular/core';

class NgxSearchPipe {
    transform(items, searchText) {
        if (items && searchText) {
            return items.filter(item => {
                let flat = {};
                const flatObject = (obj) => {
                    Object.keys(obj).forEach((key) => {
                        if (typeof obj[key] === 'object' && obj[key] !== null) {
                            const flatChildObject = flatObject(obj[key]);
                            Object.keys(flatChildObject).forEach((childKey) => {
                                flat[`${key}.${childKey}`] = flatChildObject[childKey];
                            });
                        }
                        else
                            flat[key] = obj[key];
                    });
                    return flat;
                };
                return Object.values(flatObject(item)).join(' ').toLowerCase().includes(searchText.toLowerCase());
            });
        }
        return items || [];
    }
}
NgxSearchPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NgxSearchPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
NgxSearchPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "15.2.4", ngImport: i0, type: NgxSearchPipe, name: "ngxSearchPipe" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NgxSearchPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'ngxSearchPipe'
                }]
        }] });

class NgxSearchPipeModule {
}
NgxSearchPipeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NgxSearchPipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxSearchPipeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.4", ngImport: i0, type: NgxSearchPipeModule, declarations: [NgxSearchPipe], exports: [NgxSearchPipe] });
NgxSearchPipeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NgxSearchPipeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NgxSearchPipeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxSearchPipe],
                    exports: [NgxSearchPipe]
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgxSearchPipe, NgxSearchPipeModule };
